Hosting dashboard uses the bootstrapped user object where available

p-jackson opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->


## Proposed Changes

Adds the existing `wpcom-user-bootstrap` feature to the hosting dashboard. If the node server places the user object in the `window.currentUser` global, then we should use it in the hosting dashboard to start rendering our UI faster.

This can not be done locally unfortunately, it is only possible in environments where the `wpcom-user-bootstrap` feature is enabled.

The equivalent V1 code is here.

https://github.com/Automattic/wp-calypso/blob/32599d702ce04b1c9e0de5a9c1975353cf3b9964/client/lib/user/shared-utils/initialize-current-user.ts#L29

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

The `wpcom-user-bootstrap` feature is primarily a performance optimisation.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

This one is tricky to test without merging. We can't test locally because our local calypso servers don't have a secret key needed to access fetch user data.

Please confirm that localhost still works as it did before, but the final test will need to be done on staging, and a quick revert prepared for.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

p-jackson committed:
Hosting dashboard uses the bootstrapped user object where available

p-jackson commented on client/dashboard/app/auth/index.tsx:
```diff
@@ -14,6 +15,23 @@ interface AuthContextType {
 }
 export const AuthContext = createContext< AuthContextType | undefined >( undefined );
 
+async function initializeCurrentUser(): Promise< User > {
+	// In support user session the `currentUser` refers to the wrong person so we should request
+	// the user object. Note we do not check `isSupportNextSession()` because in "next" support
+	// sessions the server does bootstrap the correct `currentUser`.
+	const useBootstrap = ! isSupportUserSession() && config.isEnabled( 'wpcom-user-bootstrap' );
+
+	if ( useBootstrap ) {
+		// eslint-disable-next-line @typescript-eslint/no-explicit-any
+		if ( ( window as any ).currentUser ) {
+			// eslint-disable-next-line @typescript-eslint/no-explicit-any
+			return ( window as any ).currentUser;
+		}
+	}
```
The logic here is slightly different from V1. In V1 if the bootstrap object _should_ be there but is not, then it redirects the user.

I don't see the harm in treating the bootstrap as an optional optimisation, and so falling back to the default behaviour of trying to fetching from public-api if the bootstrap is unavailable.

jsnajdr replied:
There will be some visible effects for logged-out users who navigate to a hosting dashboard URL:
- more `/me` requests to the server with failed authentication. There is a Grafana dashboard for this, we should see a little spike there
- the logged out user will be redirected to login page, but it will take longer than usual because there is the extra roundtrip trying to fetch `/me`.

Keep the fallback for some time if you don't feel confident changing this code, but the mainline Calypso doesn't have it 🙂

The TypeScript workarounds could be avoided by augmenting the `window` type:
```ts
declare global {
  interface Window {
    currentUser?: User;
  }
}
```
Both here and in the original Calypso code. Such `window` augmentations are done at 30+ places across Calypso, for custom globals like this.

p-jackson committed:
Remove unnecessary window type casting

p-jackson committed:
Do not attempt to fetch user if the bootstraping fails

jsnajdr commented on client/dashboard/app/auth/index.tsx:
```diff
@@ -14,6 +15,27 @@ interface AuthContextType {
+			// - client/lib/user/user.d.ts
+			return window.currentUser as any;
+		}
+		throw new Error( 'Failed to bootstrap user object' );
```
Are you sure this won't break anonymous requests? Such a request should redirect to login, not display some 404 error.

p-jackson replied:
Since user object loading is handled by ReactQuery, an exception is how we communicate to ReactQuery that the loading failed. That sets the `AuthProvider`'s `userIsError` flag, and that causes the redirect to the login page, not a generic error page.

> won't break anonymous requests

The hosting dashboard doesn't have any logged out requests (although that might change if we were create a log in interface in the hosting dashboard). The `AuthProvider` renders `null` during loading, and forces a redirect if there's no user. So it is all or nothing if we can't bootstrap this user object :D

jsnajdr replied:
> The hosting dashboard doesn't have any logged out requests

That means that all logged out requests are immediately redirected to login. If you're sure that this redirect will really happen, and not some 404 page, that's all I was worried about 🙂

p-jackson committed:
Fix error in type_check_client

The window object can not be extended because, in the editor, it
believes that the `currentUser` property has already been set (by a type
definition in `data-stores`). However if we don't extend the type, then
the type_check_client script doesn't see the `currentUser` property
definition and it errors.

This should be fixed by aligning all the various user types together.
